streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1543,26 +1544,28 @@ private boolean close(final Optional<Long> timeout, final boolean leaveGroup) {
             timeoutMs = Long.MAX_VALUE;
         }
 
-        if (state.hasCompletedShutdown()) {
-            log.info("Streams client is already in the terminal {} state, all resources are closed and the client has stopped.", state);
-            return true;
-        }
-        if (state.isShuttingDown()) {
-            log.info("Streams client is in {}, all resources are being closed and the client will be stopped.", state);
-            if (state == State.PENDING_ERROR && waitOnState(State.ERROR, timeoutMs)) {
-                log.info("Streams client stopped to ERROR completely");
-                return true;
-            } else if (state == State.PENDING_SHUTDOWN && waitOnState(State.NOT_RUNNING, timeoutMs)) {
-                log.info("Streams client stopped to NOT_RUNNING completely");
+        if (!setState(State.PENDING_SHUTDOWN)) {
+
+            if (state.isShuttingDown()) {
+                log.info("Streams client is in {}, all resources are being closed and the client will be stopped.", state);
+                if (state == State.PENDING_ERROR && waitOnState(State.ERROR, timeoutMs)) {
+                    log.info("Streams client stopped to ERROR completely");
+                    return true;
+                } else if (state == State.PENDING_SHUTDOWN && waitOnState(State.NOT_RUNNING, timeoutMs)) {
+                    log.info("Streams client stopped to NOT_RUNNING completely");
+                    return true;
+                } else {
+                    log.warn("Streams client cannot transition to {} completely within the timeout",
+                        state == State.PENDING_SHUTDOWN ? State.NOT_RUNNING : State.ERROR);
+                    return false;
+                }
+            }
+
+            if (state.hasCompletedShutdown()) {
+                log.info("Streams client is already in the terminal {} state, all resources are closed and the client has stopped.", state);
                 return true;
-            } else {
-                log.warn("Streams client cannot transition to {} completely within the timeout",
-                         state == State.PENDING_SHUTDOWN ? State.NOT_RUNNING : State.ERROR);
-                return false;
             }
-        }
 
-        if (!setState(State.PENDING_SHUTDOWN)) {
             // if we can't transition to PENDING_SHUTDOWN but not because we're already shutting down, then it must be fatal
```

Comment:
I think this condition cannot be reached any longer, given the new logic? After `if(!setState(State.PENDING_SHUTDOWN))` we know we are in one of the four (pending) terminal states already.